### ISSUE TITLE
User matching problem with Azure OIDC

### ISSUE THREAD

**Issue by ChrisCYC** [microsoft/o365-moodle#2139]
Moodle version 3.11.8
Plugin version: 2021051725

So I have  a similar issue that has been reported previously. 

We have used this plugin successfully in the past few years to match new LDAP users to their O365 accounts and switch their authentication methods to OIDC via the **Sync users with Azure AD** scheduled task. 

Historically we have always matched using email address so here is a screenshot of our sync settings over the past few years. 

![image](https://user-images.githubusercontent.com/73239085/185365870-c9eb890c-6b02-4efd-9367-6ebcb1a4bee9.png)

This means they will login with their O365 accounts from when they first login to Moodle. 

Since our plugin upgrade, despite none of the fields being amended, the users are matched, but their authentication is no longer switched from LDAP to OIDC when the match is made. 

I have tried switching between matching username and email, but the same issue occurs. 

![image](https://user-images.githubusercontent.com/73239085/185357720-2f275b1f-5d3b-473f-a7a2-5d52f4490689.png)

It looks like the plugin wants the user to confirm the change as opposed to performing the full sync automatically. 

![image](https://user-images.githubusercontent.com/73239085/185361139-9139f881-a506-4b3b-aa43-f28e9811a6c4.png)

Any ideas would be very well received!

**Comment by ChrisCYC** [comment]
Historically we have always matched using email address so here is a screenshot of our sync settings over the past few years. 

![image](https://user-images.githubusercontent.com/73239085/185365870-c9eb890c-6b02-4efd-9367-6ebcb1a4bee9.png)

**Comment by weilai-irl** [comment]
Hi @ChrisCYC,

Could you clarify a few details please
 - For the Moodle user with username 708232, what is the email address in Moodle profile?
 - What is the value in the "Minimum inexact username length to switch to Microsoft 365" setting in the "Advanced" tab on the configuration page?

For a bit of background, the only reason the sync task would show "Matched user, but did not switch them to OIDC" is when the task can't treat the match as an "exact match".
 - When processing users, the task would first create a cache of all Moodle users, and the key of the cache depends on your user sync option - if the "Match Azure usernames to moodle emails instead of moodle usernames during the sync" is checked, then the key is "email", otherwise the key is username.
 - Then for each of the Azure AD user received, the task will create three alternative versions of its upn:
    - `upnlower`: UPN converted to lower case letters.
    - `upnsplit0`: Extract the username part of the upn, i.e. strip the domain part.
    - `convertedupn`: For guest users in the Azure AD, which would have "#EXT#" in their upn, this would contain the lower case of their Azure AD account email profile. 
  - When processing a user:
    - If `upnlower` exists in the user cache keys, the user is treated as an exact match, otherwise,
    - If `upnsplit0` exists in the user cache keys,
       - If the length of `upnsplit0` string is equal to or larger than the "Minimum inexact username length to switch to Microsoft 365" setting (default to 10), then this is treated as an exact match;
       - **If the `upnsplit0` is not shorter than the setting, it's treated as not an exact match.**
    - If `convertedupn` exists in the user cache keys, it's treated as an exact match.

So effectively the only situation the user would be treated as not an exact match is:
 - `upnlower` doesn't exist in user cache keys, AND
 - `upnsplit0` exists in user cache keys, AND
 - The length of `upnsplit0` is shorter than the "Minimum inexact username length to switch to Microsoft 365" setting.

With this in mind, could you investigate why the user was treated as not exact match in your case please. You can add the following lines to https://github.com/microsoft/o365-moodle/blob/c91d801c8e55707b875a3ae85913ae709528d9bc/local/o365/classes/feature/usersync/main.php#L1080 to let the task print out the user received, and the keys in the existing user cache for comparison.

```
            if ($aaduser['userPrincipalName'] == '708232@yeovil.ac.uk') {
                var_dump($aaduser);
                var_dump(array_keys($existingusers));
            }
```

Please give it a try and post back progress.

Regards,
Lai

**Comment by ChrisCYC** [comment]
Hi Lai, thanks for getting back to me. 

- The email would be 708232@yeovil.ac.uk 
-The  'Minimum inexact username length to switch to Microsoft 365' has been left as blank (default being 10).

So by taking your above logic into account, because all our students ID numbers used for O365 username and email are less than 10 characters (6 characters) they would be treated as not an exact match because they are shorter than the minimum specified character limit. 

I will try setting this to 'Minimum inexact username length to switch to Microsoft 365'  to '5' and run the scheduled task.

**Comment by ChrisCYC** [comment]
Excellent, that has fixed it. I have checked the user in the second screenshot within the original post (709439) and the scheduled task updated their authentication method.

![image1](https://user-images.githubusercontent.com/73239085/185628130-818efa4d-b5e5-405c-923e-50a632b369df.png)

And this is reflected in the 'Manage User connections' screen. 

![image2](https://user-images.githubusercontent.com/73239085/185628256-9d4bec23-c98c-4e24-b020-118abb9b8d73.png)

The only thing I can think is that this settings wasn't created in the version of the plugin that we upgraded from. 

Many thanks for your help Lai
